File: clients/web/src/app/shared/utils/payment.util.ts
import { formatGridCellValue } from './field-display.util';
import { formatCurrency, formatDate, type LocaleTag } from './locale-format.util';

export const VENDOR_PAYMENT_PO_FIELD = 'po_id';
export const CUSTOMER_PAYMENT_INVOICE_FIELD = 'invoice_id';

export interface PaymentSummaryLabels {
  total: string;
  paid: string;
  balance: string;
}

/** Filter payment rows allocated to a parent document. */
export function filterPaymentsByParent(
  payments: Record<string, unknown>[],
  parentField: string,
  parentId: string,
): Record<string, unknown>[] {
  return payments.filter((row) => String(row[parentField] ?? '') === parentId);
}

export function filterVendorPayments(
  payments: Record<string, unknown>[],
  poId: string,
): Record<string, unknown>[] {
  return filterPaymentsByParent(payments, VENDOR_PAYMENT_PO_FIELD, poId);
}

export function filterCustomerPayments(
  payments: Record<string, unknown>[],
  invoiceId: string,
): Record<string, unknown>[] {
  return filterPaymentsByParent(payments, CUSTOMER_PAYMENT_INVOICE_FIELD, invoiceId);
}

export function paymentAmount(payment: Record<string, unknown>): number {
  const value = Number(payment['amount']);
  return Number.isFinite(value) ? value : 0;
}

export function sumPaymentAmounts(payments: Record<string, unknown>[]): number {
  return payments.reduce((sum, row) => sum + paymentAmount(row), 0);
}

export function formatPaymentAmount(
  amount: number,
  currencyCode = 'USD',
  localeTag: LocaleTag = 'en-US',
): string {
  if (amount === 0) {
    return '—';
  }
  return formatCurrency(amount, currencyCode, localeTag);
}

export function formatPaymentCell(
  payment: Record<string, unknown>,
  currencyCode = 'USD',
  localeTag: LocaleTag = 'en-US',
): string {
  return formatPaymentAmount(paymentAmount(payment), currencyCode, localeTag);
}

export function formatPaymentDate(
  payment: Record<string, unknown>,
  localeTag: LocaleTag = 'en-US',
): string {
  const raw = payment['payment_date'];
  if (raw === undefined || raw === null || raw === '') {
    return '—';
  }
  try {
    return formatDate(raw as string | Date, localeTag);
  } catch {
    return formatGridCellValue('payment_date', raw, { fieldType: 'date' });
  }
}

export function paymentNumberLabel(payment: Record<string, unknown>): string {
  const num = String(payment['payment_number'] ?? '').trim();
  return num || String(payment['id'] ?? '—');
}

export function paymentStatusLabel(payment: Record<string, unknown>): string {
  const status = String(payment['status'] ?? '').trim();
  return status || '—';
}

export function numericField(record: Record<string, unknown>, field: string): number {
  const value = Number(record[field]);
  return Number.isFinite(value) ? value : 0;
}

/** Build formatted total / paid / balance labels from a parent record. */
export function buildPaymentSummaryLabels(
  record: Record<string, unknown>,
  totalField: string,
  currencyCode = 'USD',
  localeTag: LocaleTag = 'en-US',
): PaymentSummaryLabels {
  const total = numericField(record, totalField);
  const paid = numericField(record, 'amount_paid');
  const balance = numericField(record, 'balance_due');
  return {
    total: formatPaymentAmount(total, currencyCode, localeTag),
    paid: formatPaymentAmount(paid, currencyCode, localeTag),
    balance: formatPaymentAmount(balance, currencyCode, localeTag),
  };
}

export function hasOutstandingBalance(record: Record<string, unknown>): boolean {
  return numericField(record, 'balance_due') > 0;
}

File: clients/mobile/lib/utils/payment_util.dart
import 'locale_format_util.dart';

/// Payment summary + AP/AR action helpers — web entity-record parity.

const vendorPaymentEntityCode = 'VENDOR_PAYMENT';
const customerPaymentEntityCode = 'CUSTOMER_PAYMENT';
const vendorPaymentPoField = 'po_id';
const customerPaymentInvoiceField = 'invoice_id';

class PaymentSummary {
  const PaymentSummary({
    required this.total,
    required this.paid,
    required this.balance,
  });

  final double total;
  final double paid;
  final double balance;
}

double _amount(dynamic value) {
  return double.tryParse('$value') ?? 0;
}

bool hasOutstandingBalance(Map<String, dynamic> record) {
  return _amount(record['balance_due']) > 0;
}

bool showPaymentSummaryCard(String entityCode) {
  return entityCode == 'PURCHASE_ORDER' || entityCode == 'INVOICE';
}

PaymentSummary? buildPaymentSummary(
  String entityCode,
  Map<String, dynamic> record,
) {
  if (entityCode == 'PURCHASE_ORDER') {
    final total = _amount(record['total_amount']);
    final paid = _amount(record['amount_paid']);
    final balance = record.containsKey('balance_due')
        ? _amount(record['balance_due'])
        : total - paid;
    return PaymentSummary(total: total, paid: paid, balance: balance);
  }
  if (entityCode == 'INVOICE') {
    final total = _amount(record['amount']);
    final paid = _amount(record['amount_paid']);
    final balance = record.containsKey('balance_due')
        ? _amount(record['balance_due'])
        : total - paid;
    return PaymentSummary(total: total, paid: paid, balance: balance);
  }
  return null;
}

bool canRecordVendorPayment(
  String entityCode,
  Map<String, dynamic> record, {
  String? recordId,
  bool creatingNew = false,
}) {
  final id = recordId ?? '${record['id'] ?? ''}';
  if (entityCode != 'PURCHASE_ORDER' || id.isEmpty || creatingNew) {
    return false;
  }
  if ('${record['status'] ?? ''}' == 'cancelled') {
    return false;
  }
  return hasOutstandingBalance(record);
}

bool canCollectCustomerPayment(
  String entityCode,
  Map<String, dynamic> record, {
  String? recordId,
  bool creatingNew = false,
}) {
  final id = recordId ?? '${record['id'] ?? ''}';
  if (entityCode != 'INVOICE' || id.isEmpty || creatingNew) {
    return false;
  }
  final status = '${record['status'] ?? ''}';
  if (status == 'paid' || status == 'void') {
    return false;
  }
  return hasOutstandingBalance(record);
}

Map<String, String> vendorPaymentPrefill(
  Map<String, dynamic> poRecord,
  String poId,
) {
  final prefill = <String, String>{'po_id': poId};
  final supplierId = poRecord['supplier_id'];
  if (supplierId != null && '$supplierId'.isNotEmpty) {
    prefill['supplier_id'] = '$supplierId';
  }
  return prefill;
}

Map<String, String> customerPaymentPrefill(
  Map<String, dynamic> invoiceRecord,
  String invoiceId,
) {
  final prefill = <String, String>{'invoice_id': invoiceId};
  final customerId = invoiceRecord['customer_id'];
  if (customerId != null && '$customerId'.isNotEmpty) {
    prefill['customer_id'] = '$customerId';
  }
  return prefill;
}

List<Map<String, dynamic>> filterPaymentsByParent(
  List<Map<String, dynamic>> payments,
  String parentField,
  String parentId,
) {
  return payments
      .where((row) => '${row[parentField] ?? ''}' == parentId)
      .toList();
}

List<Map<String, dynamic>> filterVendorPayments(
  List<Map<String, dynamic>> payments,
  String poId,
) {
  return filterPaymentsByParent(payments, vendorPaymentPoField, poId);
}

List<Map<String, dynamic>> filterCustomerPayments(
  List<Map<String, dynamic>> payments,
  String invoiceId,
) {
  return filterPaymentsByParent(payments, customerPaymentInvoiceField, invoiceId);
}

bool showVendorPaymentsSection(String entityCode) {
  return entityCode == 'PURCHASE_ORDER';
}

bool showCustomerPaymentsSection(String entityCode) {
  return entityCode == 'INVOICE';
}

double paymentAmount(Map<String, dynamic> payment) {
  return _amount(payment['amount']);
}

String paymentNumberLabel(Map<String, dynamic> payment) {
  final num = '${payment['payment_number'] ?? ''}'.trim();
  return num.isNotEmpty ? num : '${payment['id'] ?? '—'}';
}

String paymentStatusLabel(Map<String, dynamic> payment) {
  final status = '${payment['status'] ?? ''}'.trim();
  return status.isNotEmpty ? status : '—';
}

String formatPaymentAmount(
  double amount, {
  String currencyCode = 'USD',
  String locale = 'en-US',
}) {
  if (amount == 0) {
    return '—';
  }
  return formatCurrency(amount, currencyCode, locale);
}

String formatPaymentDate(
  Map<String, dynamic> payment, {
  String locale = 'en-US',
}) {
  final raw = payment['payment_date'];
  if (raw == null || raw == '') {
    return '—';
  }
  final parsed = DateTime.tryParse('$raw');
  if (parsed == null) {
    return '$raw';
  }
  return formatDate(parsed, locale);
}
